src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,224 @@
+    get(*packaged_char_speed_v_minus) =
+        get<2>(*packaged_char_speeds) *
+        (get(pi) - normal_dot_phi - get(*packaged_char_speed_gamma2_v_psi));
+  }
+
+  for (size_t d = 0; d < Dim; ++d) {
+    packaged_char_speed_n_times_v_plus->get(d) =
+        get(*packaged_char_speed_v_plus) * normal_covector.get(d);
+    packaged_char_speed_n_times_v_minus->get(d) =
+        get(*packaged_char_speed_v_minus) * normal_covector.get(d);
+  }
+
+  get(*packaged_char_speed_v_psi) = get<0>(*packaged_char_speeds) * get(psi);
+  get(*packaged_char_speed_gamma2_v_psi) *= get<0>(*packaged_char_speeds);
+
+  return max(max(get<0>(*packaged_char_speeds), get<1>(*packaged_char_speeds),
+                 get<2>(*packaged_char_speeds)));
+}
+
+template <size_t Dim>
+void UpwindPenalty<Dim>::dg_boundary_terms(
+    const gsl::not_null<Scalar<DataVector>*> pi_boundary_correction,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        phi_boundary_correction,
+    const gsl::not_null<Scalar<DataVector>*> psi_boundary_correction,
+
+    const Scalar<DataVector>& char_speed_v_psi_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& char_speed_v_zero_int,
+    const Scalar<DataVector>& char_speed_v_plus_int,
+    const Scalar<DataVector>& char_speed_v_minus_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        char_speed_normal_times_v_plus_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        char_speed_normal_times_v_minus_int,
+    const Scalar<DataVector>& char_speed_constraint_gamma2_v_psi_int,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& char_speeds_int,
+
+    const Scalar<DataVector>& char_speed_v_psi_ext,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& char_speed_v_zero_ext,
+    const Scalar<DataVector>& char_speed_v_plus_ext,
+    const Scalar<DataVector>& char_speed_v_minus_ext,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        char_speed_minus_normal_times_v_plus_ext,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        char_speed_minus_normal_times_v_minus_ext,
+    const Scalar<DataVector>& char_speed_constraint_gamma2_v_psi_ext,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& char_speeds_ext,
+    dg::Formulation /*dg_formulation*/) const noexcept {
+  const size_t num_pts = char_speeds_int[0].size();
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>,
+                       ::Tags::TempScalar<2>, ::Tags::TempScalar<3>,
+                       ::Tags::TempScalar<4>, ::Tags::TempScalar<5>,
+                       ::Tags::TempScalar<6>, ::Tags::TempScalar<7>>>
+      buffer(num_pts);
+  DataVector& weighted_lambda_psi_int = get(get<::Tags::TempScalar<0>>(buffer));
+  weighted_lambda_psi_int = step_function(-char_speeds_int[0]);
+  DataVector& weighted_lambda_psi_ext = get(get<::Tags::TempScalar<1>>(buffer));
+  weighted_lambda_psi_ext = -step_function(char_speeds_ext[0]);
```

Comment:
From the documentation, I anticipated some check regarding the `\f$n\f$` vectors, but here it looks like the only version implemented is the one in which `n_i^{ext}` points in the opposite direction to `n_i^{int}`? That makes sense to me, logically, but in the presentation of the documentation it would be helpful to point out that the implementation makes this assumption.